projects/solar/src/solar_fsm.c

```diff
@@ -0,0 +1,92 @@
+#include "solar_fsm.h"
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "can.h"
+#include "can_msg_defs.h"
+#include "can_unpack.h"
+#include "drv120_relay.h"
+#include "event_queue.h"
+#include "exported_enums.h"
+#include "fsm.h"
+#include "solar_events.h"
+#include "status.h"
+
+FSM_DECLARE_STATE(state_relay_open);
+FSM_DECLARE_STATE(state_relay_closed);
+
+static bool prv_relay_open_fault_guard(const Fsm *fsm, const Event *e, void *context) {
+  SolarFsmStorage *storage = context;
+  EESolarFault fault = GET_FAULT_FROM_EVENT(*e);
+  for (uint8_t i = 0; i < storage->num_relay_open_faults; i++) {
+    if (storage->relay_open_faults[i] == fault) {
+      return true;
+    }
+  }
+  return false;
+}
+
+FSM_STATE_TRANSITION(state_relay_open) {
+  FSM_ADD_TRANSITION(SOLAR_COMMAND_EVENT_CLOSE_RELAY, state_relay_closed);
+}
+
+FSM_STATE_TRANSITION(state_relay_closed) {
+  FSM_ADD_TRANSITION(SOLAR_COMMAND_EVENT_OPEN_RELAY, state_relay_open);
+  FSM_ADD_GUARDED_TRANSITION(SOLAR_FAULT_EVENT, prv_relay_open_fault_guard, state_relay_open);
+}
+
+static void prv_open_relay(Fsm *fsm, const Event *e, void *context) {
+  drv120_relay_open();
+}
+
+static void prv_close_relay(Fsm *fsm, const Event *e, void *context) {
+  drv120_relay_close();
+}
+
+static StatusCode prv_handle_relay_state_can(const CanMessage *msg, void *context,
+                                             CanAckStatus *ack) {
+  // mirror the battery relay state
+  uint16_t relay_mask, relay_state;
+  CAN_UNPACK_SET_RELAY_STATES(msg, &relay_mask, &relay_state);
+
+  if (relay_mask & (1 << EE_RELAY_ID_BATTERY)) {
+    if (relay_state & (EE_RELAY_STATE_CLOSE << EE_RELAY_ID_BATTERY)) {
+      return event_raise_no_data(SOLAR_COMMAND_EVENT_CLOSE_RELAY);
+    } else {
+      return event_raise_no_data(SOLAR_COMMAND_EVENT_OPEN_RELAY);
+    }
+  }
+
+  return STATUS_CODE_OK;
+}
```

Comment:
now that I'm seeing it, I feel like this is sufficiently different from the rest of this code for it to be separate. I don't think I'd expect to see this here based on filenames ...